Link error: multiple definition error of `associate_descriptor` when including `<cuda/annotated_ptr>`

When I try to include the header `<cuda/annotated_ptr>` in multiple compile units, the linker returns multiple definition errors indicating something gets repeatedly defined.

- source files
```cuda
// a.cu
#include <cuda/annotated_ptr>

__global__ void hello();

int main() {
  hello<<<1, 1>>>();
  return 0;
}
```

```cuda
// b.cu
#include <cstdio>
#include <cuda/annotated_ptr>

__global__ void hello() {
  printf("hello from (%d, %d)\n", blockIdx.x, threadIdx.x); 
}
```

NVIDIA GPU driver version: 470.82.01, `nvcc` version: 11.6.124 .
compilation command & errors:
```bash
nvcc a.cu b.cu -arch=sm_80 -o a.out

/tmp/tmpxft_00035cb1_00000000-18_b.o: In function `void* cuda::__4::__detail_ap::__associate_descriptor<unsigned long>(void*, unsigned long)':
tmpxft_00035cb1_00000000-10_b.cudafe1.cpp:(.text+0x16): multiple definition of `void* cuda::__4::__detail_ap::__associate_descriptor<unsigned long>(void*, unsigned long)'
/tmp/tmpxft_00035cb1_00000000-15_a.o:tmpxft_00035cb1_00000000-6_a.cudafe1.cpp:(.text+0x16): first defined here
/tmp/tmpxft_00035cb1_00000000-18_b.o: In function `void* cuda::__4::__detail_ap::__associate_descriptor<cuda::__4::access_property::shared>(void*, cuda::__4::access_property::shared)':
tmpxft_00035cb1_00000000-10_b.cudafe1.cpp:(.text+0x37): multiple definition of `void* cuda::__4::__detail_ap::__associate_descriptor<cuda::__4::access_property::shared>(void*, cuda::__4::access_property::shared)'
/tmp/tmpxft_00035cb1_00000000-15_a.o:tmpxft_00035cb1_00000000-6_a.cudafe1.cpp:(.text+0x37): first defined here
/tmp/tmpxft_00035cb1_00000000-18_b.o: In function `cuda::__4::discard_memory(void volatile*, unsigned long)':
tmpxft_00035cb1_00000000-10_b.cudafe1.cpp:(.text+0x54): multiple definition of `cuda::__4::discard_memory(void volatile*, unsigned long)'
/tmp/tmpxft_00035cb1_00000000-15_a.o:tmpxft_00035cb1_00000000-6_a.cudafe1.cpp:(.text+0x54): first defined here
collect2: error: ld returned 1 exit status
```

--- a/include/cuda/annotated_ptr
+++ b/include/cuda/annotated_ptr
@@ -170,6 +170,7 @@ void apply_access_property(const volatile void* __ptr, const _Shape __shape, acc
 #endif
 }
 
+inline
 __host__ __device__
 void discard_memory(volatile void* __ptr, std::size_t __nbytes) noexcept {
 #if __CUDA_ARCH__ >= 800

--- a/include/cuda/std/detail/__access_property
+++ b/include/cuda/std/detail/__access_property
@@ -155,6 +155,7 @@ namespace __detail_ap {
           std::uint64_t(__ap_reserved3)           << 63;
       }
 
+      inline
       __host__ __device__
       std::uint64_t __get_descriptor_non_cexpr() const noexcept { return *reinterpret_cast<const std::uint64_t*>(this); }
 
@@ -256,6 +257,7 @@ namespace __detail_ap {
           std::uint64_t(__ap_reserved2)           << 63;
       }
 
+      inline
       __host__ __device__
       std::uint64_t __get_descriptor_non_cexpr() const noexcept { return *reinterpret_cast<const std::uint64_t*>(this); }
 

--- a/include/cuda/std/detail/__annotated_ptr
+++ b/include/cuda/std/detail/__annotated_ptr
@@ -86,7 +86,7 @@ namespace __detail_ap {
   }
 
   template <>
-  __device__
+  inline __device__
   void* __associate_descriptor(void* __ptr, std::uint64_t __prop) {
 #if __CUDA_ARCH__ >= 800
     return __nv_associate_access_property(__ptr, __prop);
@@ -96,7 +96,7 @@ namespace __detail_ap {
   }
 
   template<>
-  __device__
+  inline __device__
   void* __associate_descriptor(void* __ptr, access_property::shared) {
     return __ptr;
   }
@@ -125,7 +125,7 @@ namespace __detail_ap {
       constexpr __annotated_ptr_base() noexcept = default;
       constexpr __annotated_ptr_base(__annotated_ptr_base const&) = default;
       _LIBCUDACXX_CONSTEXPR_AFTER_CXX11 __annotated_ptr_base& operator=(const __annotated_ptr_base&) = default;
-      __device__ void* __apply_prop(void* __p) const {
+      inline __device__ void* __apply_prop(void* __p) const {
         return __associate(__p, access_property::shared{});
       }
   };
@@ -138,7 +138,7 @@ namespace __detail_ap {
       constexpr __annotated_ptr_base() noexcept = default;
       constexpr __annotated_ptr_base(__annotated_ptr_base const&) = default;
       _LIBCUDACXX_CONSTEXPR_AFTER_CXX11 __annotated_ptr_base& operator=(const __annotated_ptr_base&) = default;
-      __device__ void* __apply_prop(void* __p) const {
+      inline __device__ void* __apply_prop(void* __p) const {
         return __associate(__p, access_property::global{});
       }
   };
@@ -151,7 +151,7 @@ namespace __detail_ap {
       constexpr __annotated_ptr_base() noexcept = default;
       constexpr __annotated_ptr_base(__annotated_ptr_base const&) = default;
       _LIBCUDACXX_CONSTEXPR_AFTER_CXX11 __annotated_ptr_base& operator=(const __annotated_ptr_base&) = default;
-      __device__ void* __apply_prop(void* __p) const {
+      inline __device__ void* __apply_prop(void* __p) const {
         return __associate(__p, access_property::normal{});
       }
   };
@@ -164,7 +164,7 @@ namespace __detail_ap {
       constexpr __annotated_ptr_base() noexcept = default;
       constexpr __annotated_ptr_base(__annotated_ptr_base const&) = default;
       _LIBCUDACXX_CONSTEXPR_AFTER_CXX11 __annotated_ptr_base& operator=(const __annotated_ptr_base&) = default;
-      __device__ void* __apply_prop(void* __p) const {
+      inline __device__ void* __apply_prop(void* __p) const {
         return __associate(__p, access_property::persisting{});
       }
   };
@@ -177,7 +177,7 @@ namespace __detail_ap {
       constexpr __annotated_ptr_base() noexcept = default;
       constexpr __annotated_ptr_base(__annotated_ptr_base const&) = default;
       _LIBCUDACXX_CONSTEXPR_AFTER_CXX11 __annotated_ptr_base& operator=(const __annotated_ptr_base&) = default;
-      __device__ void* __apply_prop(void* __p) const {
+      inline __device__ void* __apply_prop(void* __p) const {
         return __associate(__p, access_property::streaming{});
       }
   };
@@ -191,7 +191,7 @@ namespace __detail_ap {
       __host__ __device__ constexpr __annotated_ptr_base(std::uint64_t __property) noexcept : __prop(__property) {}
       constexpr __annotated_ptr_base(__annotated_ptr_base const&) = default;
       _LIBCUDACXX_CONSTEXPR_AFTER_CXX11 __annotated_ptr_base& operator=(const __annotated_ptr_base&) = default;
-      __device__ void* __apply_prop(void* __p) const {
+      inline __device__ void* __apply_prop(void* __p) const {
         return __associate(__p, __prop);
       }
   };

